Computer resources new page (#1148)
ihmeuw/vivarium_research: docs/source/onboarding_resources/computing/index.rst

#. You can start jobs on the cluster, simple tasks are covered in Module 3 within the IHME Learn training `Computational Infrastructure Level 1 <https://ihme.brightspace.com/d2l/home/7028>`_. 
#. More complex jobs are handled through Jobmon. While currently the Sim Science research team does not use Jobmon frequently, it might be helpful to understand this process with Module 4 within the IHME Learn training `Computational Infrastructure Level 2  <https://ihme.brightspace.com/d2l/home/7090>`_.

If you need help applying any of these trainings to a practical situation, please ask! 

.. _cluster_access_jupyter:

Accessing Juptyer or RStudio from the Cluster
---------------------------------------------

The other most common task for a Sim Science researcher on the cluster is to 
start a Juptyer session. Information on how to do this can be found on the Hub 
page below. You will also need to update your Bash files in order to complete 
this, which is covered in depth in the section :ref:`Your Bash Files <cluster_access_bash>`. 

.. todo::

>   Add central comp link to this!   

Reviewer: https://hub.ihme.washington.edu/display/DataScience/My+first+Jupyter+Notebook

Once you have started a session, you will be able to create code, test simulation 
results, or do quick calculations. Once you have finished coding, you’ll want to 
follow the same steps as outlined above in the :ref:`Contributing New Documentation <contributing>` 
section to save the information on GitHub. All of the same Git commands work on 
the terminal used for cluster access the same way as for GitHub. 

You will need to make sure that you have cloned your repository and are in the 
appropriate working directory while logged into the cluster. Then you can add, 
check the status, commit, and push information in a similar way. For many projects, 
you will create a new GitHub repository that is research team specific to not 
disturb the engineers coding workflow. 

> One difference in uploading to GitHub is that the cluster will require a password 

Reviewer: This is not unique to the cluster. The same goes for your local machine.

to push information. This is **NOT** your GitHub password, but instead is a unique 
token that you will need to create. `This website <https://techglimpse.com/git-push-github-token-based-passwordless/>`_ has information on creating a token. Many Sim Science users set their token to 
never expire and save the token where they can reference it later. However, this 
might compromise security in some cases, so regenerating a token periodically is 
best practice. 

.. _cluster_access_other:

Other Useful Cluster Tips
-------------------------

#. If you get tired of typing long commands, one option is to make an alias. An alias is a shortcut command for commonly typed things. More information on how to do was written by the `Cost Effectiveness team <https://hub.ihme.washington.edu/display/CE/Setting+up+cluster+access>`_ 

> #. When your computer falls asleep, it will stop access to the cluster and cut off any jobs that were currently running. This can be problematic if a notebook needs to run overnight. There are a few different options to account for this including: screen, MOSH, or tmux. If you need to use these, ask a teammate. 

Reviewer: ```suggestion
#. When your computer falls asleep, it will stop access to the cluster and cut off any interactive jobs (i.e. :code:`srun` sessions) that were currently running. This can be problematic if a command needs to run overnight. There are a few different options to account for this including: screen, MOSH, or tmux. If you need to use these, ask a teammate. 
```

This shouldn't kill non-interactive jobs; if you use the recommended scripts to start Jupyter or RStudio, they won't be interactive jobs, so they won't be killed if you get disconnected.

.. _cluster_access_files:

File Systems and Storage 
------------------------

The cluster can be confusing with where to store code and data. Our 
team has created some best practices to use for data storage. 

For code, please create a new directory under :code:`/ihme/code` with your 
username. For example, this might be :code:`/ihme/code/lutzes`. 
You should clone GitHub respositories to this location and have 
all Jupyter notebooks and other code stored here. 

For data files, there are two locations based on the size of the 
data file. 

> #. For small data files, store these on GitHub in the same location as your code. A "small" file is something without thousands of rows of data. Examples might include: a list of nicknames, disease severity proportions by age/sex group, or drug efficacy data. 

Reviewer: ```suggestion
#. For small data files, store these on GitHub in the same location as your code. Examples might include: a list of nicknames, disease severity proportions by age/sex group, or drug efficacy data. The absolute maximum file size on GitHub is 100 megabytes, but be mindful of including any file over 10 megabytes, especially if there are many such files or if the file changes frequently. Too many large files can slow down the process of making new clones of the repository.
```

Maybe we shouldn't delete, but you can definitely get away with a few thousand rows depending on the columns etc!

#. For large files, store these in a shared location on the cluster. Considering making a new folder for each project for data storage. 

> #. When you deciding where to store data, please also consider any data restrictions that might exist. 

Reviewer: ```suggestion
#. When you decide where to store data, please also consider any data access or privacy restrictions that might exist. 
```

Regardless of where you store data, it is important to track updates 
to data files carefully. Engineers might copy and paste a file into a 
new location, so updating the file might not actually change what is 
being used in the sim. Therefore, follow these steps: 

#. Use the naming conventions below to ensure consistency. 

> #. Always version up rather than replacing a data file (e.g., create a new file with the current date rather than just replacing with a different file of the same name). 

Reviewer: ```suggestion
#. Always version up rather than replacing a data file that is used by engineering or is not tracked in GitHub (e.g., create a new file with the current date rather than just replacing with a different file of the same name). 
```

#. Include the exact file name and location in the docs. This means if you version up a data file, you will need to update the docs to reflect the new name. This ensures the engineers are aware of any changes. 

For consistency, please use this naming convention for all files: :code:`FILENAME_20230309.ext`. 
For example, this might be :code:`heart_failure_proportions_20230310.csv` 

.. _cluster_access_bash:

> Your Bash Files 

Reviewer: As noted above, these are bash *configuration* files.

---------------

Bash files contain commands you would enter into a command line, but 
specifically ones you will use very often. For example, every time you 
want to open a Juptyer session, the cluster needs certain information 
and requirements. To find this information, it looks in your Bash files 

> rather than asking you to enter the same information every time. 

Reviewer: This is a bit inaccurate and potentially confusing. `.bash_profile` and `.bashrc` are scripts that run commands every time you start a new shell. It's by creating **aliases and functions** in these files that you are able to "save commands for later" by giving a short name to a common command.

Is there not a Hub page already that explains these files? I searched (useless, of course) and didn't find one. Happy to help write this part or look for non-IHME resources if we don't find anything. The Hub page linked below gives examples but does not explain how they work (and it seems like almost all of the examples don't apply to Sim Science).

Author: I just added this as a to-do for now since this PR is already giant and overwhelming!

However, Bash files can be confusing since it is less obvious when the 
information is being used or what it is used for. Therefore, we have 
provided a copy-and-paste formatting for information to be added to 
your Bash files. If you would like more information and what this 

> code means and why we need it, this `Hub page <https://hub.ihme.washington.edu/pages/viewpage.action?pageId=83205636#UsingtheCluster&Bash-Bash>`_ has a number of helpful links. 

Reviewer: Not sure this Hub page adds much. It lists the files under "What you should know" but doesn't explain them.

.. todo::

>   Confirm and paste in bash files 

Reviewer: Let's chat to see whether there is anything most people on Sim Science use/need in these files!

.. _conda_environments:

3.0 Conda Environments
++++++++++++++++++++++

> Your environment is a “space” in which you can run code that has some 
> preset information. Environments are helpful as you can preinstall packages 
> into them for easier ability to run programs. GBD has a common environment 
> that you can use, but usually each project will also create its own environment 
> as well. Below are some common questions on environments. 

Reviewer: ```suggestion
A conda environment is a "workspace" in which you can run code with certain packages installed.
You can install a package in a conda environment without affecting any other conda
environments: they are **isolated** from one another.
This allows you to have multiple projects that each have their own separate set of packages
and package versions.
The Central Computation team maintains an environment, which anyone at IHME can use,
that includes all the packages necessary for accessing GBD results (plus some other common packages).
However, this environment is read-only; if you want any packages not included there, you will need to
make your own environment.
Generally, multiple researchers do not share the same environment: each researcher generally makes an
environment for each project they work on.
They may even make multiple if they want to use different versions in different parts of a project.
Below are some common questions on environments. 
```

**What is an environment again?**

> It’s a directory that contains a specific collection of conda packages that 

Reviewer: ```suggestion
It’s a "workspace" that contains a specific collection of packages that 
```

It's unnecessarily complicated to get into here, but the conda environment is also holding pip packages, not only conda packages.

you have installed. Basically, it is a shortcut to have all the relevant 
packages you need for a project in one place. 

**What are the advantages to having separate environments?**

> Over time, Python versioning updates with new information and bug fixes. 

Reviewer: ```suggestion
Over time, new versions of packages come out. 
```

It can therefore be helpful to create new environments to ensure you have 

> the latest python package versions. 

Reviewer: ```suggestion
the latest package versions. 
```

Do you think it's overly confusing to omit "Python" here? Conda can install non-Python packages.

While you can uninstall and reinstall new versions of packages in existing 
environments, this can sometimes cause errors in existing code. Therefore, 

> it is helpful to have code exist within environments that have correct 
> versioning and to create new environments with new projects and install the 
> latest versions of packages. 

Reviewer: ```suggestion
it is helpful to keep environments that work with existing code
and to create new environments for new projects and install the
latest versions of packages in those.
```

**How and when should I make a new one?**

> In general, our team has a new environment for each project. When you are 

Reviewer: ```suggestion
In general, our team has a new environment for each researcher on each project. When you are 
```

starting a new project, you can see if the engineering team has set up an 

> environment for that work. If not, it is best practice to consider setting one up. 

Reviewer: Wait really? Do others frequently use engineering-created environments? How does this work -- is it read-only to you?

Maybe my last comment doesn't apply then.

**How do you make a new environment?** 

.. todo::

>   Find instructions on the vivarium CVD page and include here. Also note that you need to install conda first. 

Reviewer: Since you need to install conda, or decide to use central comp's conda install, before you can even use environments, I think instructions for setting up conda should go outside this section. For example it would be needed to run a notebook as well.

**How do I install new information to an existing environment?**
Once you have made a new environment, you can add some commonly used packages 

> using :code:`-pip install package`. Common ones include numpy, pandas, or 

Reviewer: ```suggestion
using :code:`pip install package`. Common ones include numpy, pandas, or 
```

matplotlib among many others. If :code: `import` in Python fails, try 
installing the package to the environment and reloading the page. 

**When should I use the GBD environment vs a project specific one?**
In general, it is best practice to use your project’s environment for project 
work. Since this is the same environment used by the engineers on that project 

> it ensures consistency. You can use the GBD environment if you choose and it 

Reviewer: Whoa, had not heard of this! This definitely isn't the only way it is done. Some of my comments above may have been inaccurately assuming everyone did things the way I have done it.

can be helpful for some more general tasks. 

**I installed a package to this environment on the cluster - why won't it work?** 
Your local machine and the cluster are different and don’t "speak" between environments. 

> So if you install a package to an environment while on the cluster, it might not 

Reviewer: ```suggestion
So if you install a package to an environment while on the cluster, it won't
```

show on your local machine. 

**What is Python vs Conda Vs Anaconda?**
Python is the name of a programming language. It is the name for the syntax 
used in code. 

> Anaconda is a software you can use to interface with Python. It is specifically 

Reviewer: ```suggestion
Anaconda is a software you can use to install Python and conda, and create conda environments, on Windows. It is specifically 
```

Let's put this one last as in the question header! Also I'm not 100% sure I got this correct since I don't use Anaconda but I think it's roughly right.

designed for improved user interface and has tools tailored for data science. 

> Conda is package manager that we use to create and maintain environments. It is 

Reviewer: ```suggestion
Conda is a package manager that we use to create and maintain environments. It is 
```
